SA/design_concept/content/design/design_specification_sum_type_destruction.typ

```diff
@@ -0,0 +1,88 @@
+Pattern matching allows to match values against a set of patterns and
+execute a corresponding code. One of the patterns-types usually 
+destructs sum type constructors. 
+
+An example of pattern-matching in Haskell can be seen in
+@pattern-matching-example-haskell, where first a sum type `Expression` is
+defined with two constructors `Value :: Int -> Expression` and
+`Addition :: Expression -> Expression -> Expression`, then a function `calc`
+is defined that pattern-matches a value of type `Expression` against its
+constructors.
+
+#figure(
+    ```hs
+    data Expression where
+      Value :: Int -> Expression
+      Addition :: Expression -> Expression -> Expression
+
+    calc :: Expression -> Int
+    calc e = case e of
+      Value v -> v
+      Addition left right -> calc left + calc right
+    ```,
+    caption: "Example of pattern-matching in Haskell"
+)<pattern-matching-example-haskell>
+
+A more straightforward approach is to generate a function, which, for each
+constructor, takes a parameter that takes the constructor arguments as arguments
+of itself and maps it to a common value. Such an approach is called _destruction
+function_. How such a destruction function would look for the sum type
+`Expression` example can be seen in @destruction-function-example-haskell.
+
+#figure(
+    ```hs
+    data Expression where
+      Value :: Int -> Expression
+      Addition :: Expression -> Expression -> Expression
+
+    destruct :: Expression -> (Int -> b) -> (Expression -> Expression -> b) -> b
+    destruct = -- implementation omitted
+
+    calc :: Expression -> Int
+    calc e = destruct e (\x -> x) (\l r -> calc l + calc r)
+    ```,
+    caption: "Example of a destruction function in Haskell"
+)<destruction-function-example-haskell>
+
+Unfortunately, such a destruction function is not as powerful as
+pattern-matching in a few ways, for example:
+
+- It is not possible to specify multiple overlapping patterns, which are
+  matched against in order of definition.
+- It is not possible to combine patterns patterns into more complex
+  patterns.
+- It is not possible to specify a default case.
+
+This list is not exhaustive, but it already shows how the lack of
+pattern-matching would make the import of Haskell code into VisualFP more
+difficult.
+
+How these approaches would look in VisualFP can be seen in
+@pattern-matching-example-visualfp and
+@destruction-function-example-visualfp.
+
+#figure(
+  load-drawio-svg("../../static/design_specification_destruction.svg", width: 100%),
+  caption: [Example of a destruction function in VisualFP])<pattern-matching-example-visualfp>
+#figure(
+  load-drawio-svg("../../static/design_specification_pattern-matching.svg", width: 77%),
+  caption: [Example of pattern-matching in VisualFP])<destruction-function-example-visualfp>
+
+The decision was made in favor of the destruction function since the
+pattern-matching approach does not translate as well into a visual language.
```

Comment:
```suggestion
pattern-matching approach does not translate as well into a consistent visual language.
```

Reply:
What are you trying to say with 'consistent'?

Reply:
I don't find the visualization of a pattern-matching block that bad, but it doesn't fit as well to the visual language as the destruction function, in the sense that the destruction function and the other blocks have the same look & feel -> consistent visual language. Maybe there's a better word we can use instead.

Reply:
reworded